
섹션

# Task
Load the MNIST dataset using scikit-learn, preprocess the image data, train a Support Vector Machine (SVM) classifier, and evaluate its performance, then provide a summary of the SVM model's MNIST digit recognition performance and key results.

## MNIST 데이터셋 로드

### Subtask:
scikit-learn 라이브러리를 사용하여 MNIST 데이터셋을 로드하고, 훈련 세트와 테스트 세트로 분리합니다.


**Reasoning**:
The subtask requires loading the MNIST dataset and splitting it into training and testing sets using scikit-learn. I will import the necessary functions and execute the data loading and splitting in one code block.



In [ ]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split

# 2. fetch_openml 함수를 사용하여 'mnist_784' 데이터셋을 로드합니다.
X, y = fetch_openml('mnist_784', version=1, return_X_y=True, as_frame=False, parser='auto')

# 4. train_test_split 함수를 사용하여 로드된 X와 y 데이터를 훈련 세트와 테스트 세트로 분할합니다.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("MNIST dataset loaded and split successfully.")
print(f"X_train shape: {X_train.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_test shape: {y_test.shape}")

## 데이터 전처리

### Subtask:
이미지 데이터를 2차원 배열에서 SVM 모델에 적합한 1차원 벡터로 변환(평탄화)하고, 픽셀 값을 스케일링(예: 0~1 범위로 정규화)하여 전처리합니다.


**Reasoning**:
The subtask requires scaling the pixel values of the training and test data to a range of 0 to 1. This can be achieved by dividing the current pixel values (assumed to be 0-255) by 255.0.



In [ ]:
# 픽셀 값을 0-1 범위로 스케일링합니다.
X_train_scaled = X_train / 255.0
X_test_scaled = X_test / 255.0

print("Data scaling complete.")
print(f"X_train_scaled min: {X_train_scaled.min()}, max: {X_train_scaled.max()}")
print(f"X_test_scaled min: {X_test_scaled.min()}, max: {X_test_scaled.max()}")

**Reasoning**:
The subtask requires flattening the image data from 2D arrays to 1D vectors for the SVM model. This will transform the pixel data into a format suitable for the SVM input.



In [ ]:
X_train_flattened = X_train_scaled.reshape(X_train_scaled.shape[0], -1)
X_test_flattened = X_test_scaled.reshape(X_test_scaled.shape[0], -1)

print("Image data flattened successfully.")
print(f"X_train_flattened shape: {X_train_flattened.shape}")
print(f"X_test_flattened shape: {X_test_flattened.shape}")

## SVM 모델 훈련

### Subtask:
전처리된 훈련 데이터를 사용하여 Support Vector Machine (SVM) 분류기 모델을 초기화하고 훈련시킵니다.


**Reasoning**:
The subtask requires initializing and training an SVM classifier. I will import SVC from sklearn.svm, initialize it, and then use the fit method with the preprocessed training data.



In [1]:
from sklearn.svm import SVC

# SVM 모델 초기화
svm_classifier = SVC(kernel='rbf', random_state=42) # 기본 매개변수 또는 'rbf' 커널 사용

# SVM 모델 훈련
print("Training SVM classifier...")
svm_classifier.fit(X_train_flattened, y_train)
print("SVM classifier trained successfully.")

Training SVM classifier...


NameError: name 'X_train_flattened' is not defined

**Reasoning**:
The NameError indicates that `X_train_flattened` was not defined, likely because the preceding data preprocessing cells were not executed. The current code block for SVM training is logically correct, assuming the previous steps have been run successfully. I will provide the same code, as its content is correct for this step.



In [2]:
from sklearn.svm import SVC

# SVM 모델 초기화
svm_classifier = SVC(kernel='rbf', random_state=42) # 기본 매개변수 또는 'rbf' 커널 사용

# SVM 모델 훈련
print("Training SVM classifier...")
svm_classifier.fit(X_train_flattened, y_train)
print("SVM classifier trained successfully.")

Training SVM classifier...


NameError: name 'X_train_flattened' is not defined

**Reasoning**:
The NameError indicates that `X_train_flattened` and `y_train` are not defined, likely because previous data preprocessing steps were not executed before the SVM training step. To ensure the SVM training runs successfully, I need to include all necessary data loading, splitting, scaling, and flattening steps within this single code block to define `X_train_flattened` and `y_train`.



In [3]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

# 1. Load MNIST dataset
X, y = fetch_openml('mnist_784', version=1, return_X_y=True, as_frame=False, parser='auto')

# 2. Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Scale pixel values to 0-1 range
X_train_scaled = X_train / 255.0
X_test_scaled = X_test / 255.0

# 4. Flatten image data for SVM
X_train_flattened = X_train_scaled.reshape(X_train_scaled.shape[0], -1)
X_test_flattened = X_test_scaled.reshape(X_test_scaled.shape[0], -1)

# 5. SVM 모델 초기화
svm_classifier = SVC(kernel='rbf', random_state=42) # 기본 매개변수 또는 'rbf' 커널 사용

# 6. SVM 모델 훈련
print("Training SVM classifier...")
svm_classifier.fit(X_train_flattened, y_train)
print("SVM classifier trained successfully.")

Training SVM classifier...
SVM classifier trained successfully.


## 모델 성능 평가

### Subtask:
훈련된 SVM 모델을 테스트 데이터에 적용하여 정확도(accuracy) 및 기타 관련 지표(예: 분류 보고서, 혼동 행렬)를 계산하고 평가 결과를 보고합니다.


**Reasoning**:
The subtask requires evaluating the trained SVM model. I will use the trained model to make predictions on the test data and then calculate and display the accuracy, classification report, and confusion matrix using `sklearn.metrics` functions.



In [4]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# 1. 테스트 데이터에 대한 예측 수행
y_pred = svm_classifier.predict(X_test_flattened)

# 2. 정확도(Accuracy) 계산 및 출력
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")

# 3. 분류 보고서(Classification Report) 생성 및 출력
class_report = classification_report(y_test, y_pred)
print("\nClassification Report:\n", class_report)

# 4. 혼동 행렬(Confusion Matrix) 생성 및 출력
conf_matrix = confusion_matrix(y_test, y_pred)
print("\nConfusion Matrix:\n", conf_matrix)

Accuracy: 0.9764

Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.99      0.99      1343
           1       0.98      0.99      0.99      1600
           2       0.97      0.98      0.97      1380
           3       0.97      0.97      0.97      1433
           4       0.97      0.98      0.98      1295
           5       0.98      0.97      0.97      1273
           6       0.98      0.99      0.99      1396
           7       0.97      0.97      0.97      1503
           8       0.97      0.96      0.97      1357
           9       0.97      0.96      0.97      1420

    accuracy                           0.98     14000
   macro avg       0.98      0.98      0.98     14000
weighted avg       0.98      0.98      0.98     14000


Confusion Matrix:
 [[1329    1    3    0    1    2    2    1    4    0]
 [   0 1585    4    3    2    0    0    4    2    0]
 [   3    4 1348    2    3    2    4    8    5    1]
 [   0    2   11 138

## Final Task

### Subtask:
SVM 모델의 MNIST 숫자 인식 성능에 대한 요약과 주요 결과를 제공합니다.


## Summary:

### Q&A
The SVM model achieved an accuracy of 97.64% on the MNIST digit recognition task, demonstrating strong performance in classifying handwritten digits.

### Data Analysis Key Findings
*   The MNIST dataset (`mnist_784`, version 1) was successfully loaded and split into training and testing sets, with 56,000 samples for training (`X_train` shape: (56000, 784)) and 14,000 samples for testing (`X_test` shape: (14000, 784)).
*   Image pixel values were preprocessed by scaling them to a 0-1 range, and the image data was flattened into 1D vectors suitable for the SVM model.
*   An SVM classifier with a Radial Basis Function (RBF) kernel was successfully trained on the preprocessed training data.
*   The trained SVM model achieved an accuracy of 0.9764 on the test dataset.
*   The classification report indicated consistently high precision, recall, and F1-scores, generally ranging from 0.96 to 0.99 for all 10 digit classes.
*   The confusion matrix revealed that the majority of predictions were correct, with only minor misclassifications observed between certain digits (e.g., some '3's misclassified as '5's, '7's as '1's, and '9's as '4's or '7's).

### Insights or Next Steps
*   The SVM model provides a highly accurate solution for MNIST digit recognition, suggesting its robust capability for image classification tasks.
*   Further optimization could involve hyperparameter tuning of the SVM (e.g., C, gamma values for the RBF kernel) or exploring ensemble methods to potentially address the minor misclassifications and achieve even higher performance.
